experimental: ServiceCache and ServiceManager adapters

joamaki opened the pull request:
DRAFT

joamaki committed:
experimental: Support for local redirect policies

This is a port of the CiliumLocalRedirectPolicy handling to the
experimental control-plane.

This initial implementation is missing the following features:

- 'SkipRedirectFromBackend' flag / SkipLBMap
- tracking of netnscookie (for the skip functionality)
- adapting the "cilium-dbg lrp" command to dump Table[LocalRedirectPolicy]

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
ciliumenvoyconfig: Do not run reflector if envoy is disabled

Do not run the Table[CEC] reflector if the feature is disabled.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Differentiate between TCP and UDP

Now that the service TCP/UDP differentiation PR has merged, update the
experimental control-plane to use proper protocol instead of ANY.

If the experimental control-plane is used when upgrading from a cilium-agent
without the TCP/UDP differentiation the service and backend IDs are not reused
due to protocol mismatch. This should not cause traffic disruption as the backends
with ANY protocol are only replaced when pruned.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Truncate backend listing in frontend table output

Don't show more than 5 backends when showing the frontend table
in the table format to avoid making the output unreadable when
frontend has a lot of associated backends.

The full associated backends can be still inspected using yaml
or json output.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Don't show shadows without instances

In the backend output don't show a service without instances.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Flip it on by default

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Ignore object not found errors in reflector

Because of the buffering of events we may end up coalescing an
initial upsert and subsequent delete into just the delete event,
and thus we would not find the object from the tables. Since this
is expected, check for [ErrObjectNotFound] when deleting from the
LB tables.

While at it, let's switch from the panics to error logging for the
"impossible" situations.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
agent: Prepare ServiceCache and ServiceManager for adaptation

Add ServiceCache interface to make it possible to replace the implementation
with the experimental adapter. Simplify the types to the subset used by callers
so that full conversion to k8s.Service is not necessary.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

brb committed:
experimental/lb: Propagate service sessionAffinity settings

BEID=42 will become human-readable after merging
https://github.com/cilium/cilium/pull/36697.

Signed-off-by: Martynas Pumputis <martynas@isovalent.com>

joamaki committed:
experimental: Support headless services

Process headless services instead of ignoring them. Headless services should
still be synced towards Envoy.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Use terminating backends if no active are available

In order to support https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1669-proxy-terminating-endpoints
use the terminating backends as the active backends if no active backends are available.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Enable session affinity for clusterip.txtar

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Test forcing EnableSessionAffinity

EnableSessionAffinity is set by initKubeProxyReplacementOptions,
but that's probably happening after newExternalConfig() reads the
DaemonConfig. Test this out.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
TEMP: vendor bump statedb

Temporarily bump StateDB to unreleased version. This gets us statedb dump
in sysdump that's not incomplete.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: use bytes.Buffer for building BackendInstanceKey

Construct the BackendInstanceKey using bytes.Buffer to avoid
unnecessary allocations.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
ciliumenvoyconfig: Support headless services in experimental implementation

In order to support headless services we need to watch the backend table for
changes instead of frontends (since there are none for headless services).

This creates a new problem, namely that now we don't know how a port name and
port number relate, which in turn is needed if Ports are specified in the
CiliumEnvoyConfig. To solve this, add PortNames to [experimental.Service] to
allow correlating a backend with a PortName to a service port number.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Fix Frontend marshaling

Make [experimental.Frontend] marshalable by wrapping the backends iter.Seq2
type and providing MarshalJSON and MarshalYAML implementations.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
ciliumenvoyconfig: Rewrite experimental implementation

This rewrites the CEC controller and reconciler for the experimental
load-balancing control-plane. The previous version had issues handling
headless services.

The new version is implemented as a straightforward computation that reads
from Table[CEC], Table[Service] and Table[Backend] and writes the desired
state to Table[EnvoyResource]. It collects StateDB watch channels when computing
and waits for any of them to close in order to recompute.

The implementation is not terribly efficient since any change causes everything
to be recomputed. It can be made efficient by keeping separate WatchSet per
CEC and using revisions to avoid recomputing. For now I'm striving for correctness
before optimizing this.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

brb committed:
experimental/lb: Fix obsolete svc slots

We need to rely on terminating backends if there are no active backends.

Signed-off-by: Martynas Pumputis <martynas@isovalent.com>

joamaki committed:
experimental: Also create the NodePort frontends for LoadBalancer

The NodePort frontends should also be created when the service has
type LoadBalancer.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Resilient parsing of IP families, check NodePort L7 redirect

Be more resilient in parsing the service IPFamilies. Extend the
CiliumEnvoyConfig test to also check that NodePort frontends are
redirected if the ServicePort matches.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
ciliumenvoyconfig: (experimental) handle missing backend PortName

Backends without a PortName were not correctly matched up with
BackendServices that included frontend port filtering.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

brb committed:
experimental/lb: Make healthserver to bind to 0.0.0.0

Otherwise, $NODE_IP:$HEALTHSERVER_NODEPORT_PORT fails (discovered with
"[sig-network\] Services should fail health check node port if there are
only terminating endpoints" test).

Keep the old heuristiscs for the testscript tests to avoid flakes.

Signed-off-by: Martynas Pumputis <martynas@isovalent.com>

brb committed:
gh/workflows: Create LB IPAM pool in conformance-k8s tests

Otherwise, the "[sig-network] Services should fail health check node
port if there are only terminating endpoints" fails.

In the exp/lb we create HealthCheckNodePort only after a corresponding
LoadBalancer service got assigned VIP. Previously, Cilium was unable to
allocate LB VIPs.

Signed-off-by: Martynas Pumputis <martynas@isovalent.com>

brb committed:
experimental/lb: Fix healthserver tests

TODO: squash with current~2 commit.

Signed-off-by: Martynas Pumputis <martynas@isovalent.com>

brb committed:
SQUASH: gh/workflows: Fix yaml of LB IPPool

Signed-off-by: Martynas Pumputis <martynas@isovalent.com>

joamaki committed:
experimental: Do not remove backend instances on service delete

The lifetime of the backends and their references to services is
independent from the service and thus service deletion should not
influence the backend instance.

E.g. we might see a service being deleted and then added back without
the endpoint slice going away. In that scenario we shouldn't lose the
backend instances for the service. And in other data sources besides
k8s we may well insert backends very independently of the services
and frontends.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

brb committed:
experimental/lb: Use nodeport addr for HC NP backend

Otherwise, $LB_VIP:$HC_NODEPORT from outside will fail.

Signed-off-by: Martynas Pumputis <martynas@isovalent.com>

joamaki committed:
experimental: No frontends if headless annotation is set

If the service.kubernetes.io/headless annotation is set no
frontends should be created, regardless whether the service
spec would include them.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki commented on pkg/loadbalancer/experimental/healthserver.go:
```diff
@@ -183,13 +186,30 @@ func (s *healthServer) controlLoop(ctx context.Context, health cell.Health) erro
+			if !is4 {
+				beAddr = netip.IPv6Unspecified()
+			}
+			for addr := range s.params.NodeAddresses.List(wtxn, tables.NodeAddressNodePortIndex.Query(true)) {
```
This won't work since the address we pick may go away. We would need to either reconcile the changes or we could go via the NodeIP in the local node (via LocalNodeStore) as that's somewhat guaranteed not to change. Or if we can make the loopback address work that'd be even better.

brb replied:
Note to myself: ideally use https://github.com/cilium/cilium/pull/32144

brb committed:
ginkgo: Remove preflight check for |LBMAP| == |service list|

The check no longer makes sense, as "cilium service list" no longer
returns all service frontends, while "cilium bpf lb list" does.

Signed-off-by: Martynas Pumputis <martynas@isovalent.com>

joamaki committed:
experimental: Tweaks to ClusterMesh in adapter

Don't set Cluster (unless optClusterAware set).
Use the proper source.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

brb committed:
SQUASH: pkg/ciliumenvoyconfig/exp_reflector.go

Signed-off-by: Martynas Pumputis <martynas@isovalent.com>

joamaki committed:
experimental: Fix handling of dualstack NodePort frontends

The used node port addresses were key'd only by port which caused mixup with
dualstack use due to having both :: and 0.0.0.0 surrogates. Use (family, port)
as the key.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Add tests for different queries

Rename the index names to be sensible and add a test for different queries
that can be made against the service/frontend/backend tables.

Add support for querying address by prefix toi L3n4AddrFromString,
e.g. parsing "10.1.0.0/16" now returns the partial key "...\10\1".

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Add watch.sh for running tests

Add a inotifywait-based loop to rerun test when the txtar
file changes.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
ciliumenvoyconfig: Drop count= from envoy command output

The new implementation cannot guarantee whether we get one or two Envoy updates.
Fix the flakyness for now by removing the count= output.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Handle multiple port names for single backend

Handle having multiple port names for the same backend address.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Respect EnableIPv4 and EnableIPv6

Avoid unnecessarily creating BPF maps for IPv4/IPv6 when they're disabled
and avoid reconciling IPv4/IPv6 when EnableIPv4 or EnableIPv6 is disabled.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Add default maglev weight to HostPort backends

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
lbmap: Log recreation of maglev maps

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Add a test case for ExternalClusterIP

Add a test to make sure --bpf-lb-external-clusterip is correctly
implemented.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Handle HostPort handover when events overlap

If a pod with a HostPort in a deployment is restarted we may
see the events for the old and new pod overlapping and the
deletion of the pod may occur after the new pod has finished
starting. We need to make sure that the HostPort services set
up for the new pod are not messed with due to events from the
old pod.

We fix this by considering non-running pod phases being a deletion
for the HostPort. Additionally checking that a pod that has been
marked for deletion will not be processed for HostPort updates.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental/redirectpolicy: Include cell

Oops, forgot to actually reference the new implementation.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Register table initializer for adapter

In order to not prune backends before ClusterMesh has upserted them
we'll need to register an initializer and mark it complete from
SyncWithK8sFinished.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
daemon: Do not initialize maglev maps if exp lb enabled

The pkg/maps/lbmap/maglev.go will recreate the maps when table size
doesn't match, causing the map opened by pkg/loadbalancer/experimental
to be abandoned.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Provide LBMaps via bpf.MapOut to ensure ordering

Use bpf.MapOut for LBMaps to make sure they are pinned before
the loader tries to load programs.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Clean up maglev map init

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Run health server also for NodePort services

The ginkgo tests are checking for this.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Use GetLBAlgorithmAnnotation()

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
k8s/experimental: Fixes to MinimalEndpoints/MinimalService

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
exp/redirectpolicy: Fix handling of address matching

The address matching shouldn't have been trying to find an existing frontend
to redirect, but rather should create the frontend itself.

Change [Writer] to not allow overwriting frontends owned by other services.
This now requires that in the hostport test we process the pod termination
before reusing the HostPort.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

brb committed:
experimental: Delete not used maps upon startup

Signed-off-by: Martynas Pumputis <martynas@isovalent.com>

joamaki committed:
experimental: Fixes to Notifications() in adapter

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Bind to ":<port>" for HealthServer

In order to support both IPv4 and IPv6 with the HealthServer,
bind to ":<port>" rather than "0.0.0.0:<port>".

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Fix local backend lookup for health server

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Use goleak in tests

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Ignore frontends with ScopeInternal for HealthServer

We can have NodePort frontends with internal and external scope. We should
only process the one with the external scope.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Add stress.sh

Add a script for running the tests under 'stress'.

Fix up small flakyness in the healthserver.txtar.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Extended cleanup checks in nodeport.txtar

Check cleanup of orphaned backends in nodeport.txtar.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Implement adapters for ServiceCache/ServiceManager

The adapters bridge the existing ServiceCache and ServiceManager interfaces
to the new implementation, allowing us to flip over to the new implementation
without having to convert over all uses (and having double implementations).

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Use Service().Name not ServiceName when getting instance

A redirected frontend will have backends gathered from the pseudo-Service
which has a different name than [Frontend.ServiceName]. Use Service().Name
to look up backend instances so that the redirect service is used.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Fix benchmark privileged tests

BPFLBMaps wasn't set up correctly in the benchmark

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Honor EnableNodePort

Do not process NodePort/LoadBalancer/ExternalIP service types if EnableNodePort
is disabled. With kube-proxy replacement disabled this will make sure we don't
e.g. create health servers that compete with kube-proxy.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
vendor: Bump StateDB to latest unreleased version

For the WatchSet changes

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
ciliumenvoyconfig: Use the WatchSets to figure out what to recompute

Instead of using the revisions (which requires full iteration), use
the statedb.WatchSet to check if a recomputation of the Envoy resource
is required.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
ciliumenvoyconfig: Add metrics for the controller

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Do not set service when redirecting

It gets complicated quickly if we change the service reference of a frontend
to the pseudo-service when a frontend is redirected. Let's keep things cleaner
and use the "RedirectTo" solely for changing how backends are looked up.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Merge BackendInstance and BackendParams

Simplify the implementation by merging BackendInstance and BackendParams into
a single type. Use this in [Frontend.Backends] so that we don't need to do GetInstance()
when iterating over the selected backends. This also reduces confusion with local redirect
policies when the backends are for a different service name.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
vendor: Bump hive to latest version

Brings in the improvements to retry log output and the -scripttest.break

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental/redirectpolicy: Add SkipLBMap support

Signed-off-by: Jussi Maki <jussi@isovalent.com>

joamaki committed:
experimental: Benchmark for Service and Endpoint conversions

Signed-off-by: Jussi Maki <jussi@isovalent.com>
